modulus/models/gnn_layers/distributed_graph.py

```diff
@@ -223,6 +254,341 @@ def partition_graph_nodewise(
+    global_offsets: torch.Tensor,
+    global_indices: torch.Tensor,
+    partition_size: int,
+    partition_rank: int,
+    device: torch.device,
+) -> GraphPartition:  # pragma: no cover
+    """
+    Utility function which partitions a global graph given as CSC structure naively
+    by splitting both the IDs of source and destination nodes into chunks of equal
+    size. Each partition rank then manages its according chunk of both source and
+    destination nodes. Indices are assigned to the rank such that each rank manages
+    all the incoming edges for all the destination nodes on the corresponding
+    partition rank.
+    The function performs the partitioning based on a global graph in CPU
+    memory for each rank independently. It could be rewritten to e.g. only
+    do it one rank and exchange the partitions or to an algorithm that also
+    assumes an already distributed global graph, however, we expect global
+    graphs to fit in CPU memory. After the partitioning, we can get rid off
+    the larger one in CPU memory, only keep the local graphs on each GPU, and
+    avoid tedious gather/scatter routines for exchanging partitions in the process.
+
+    Parameters
+    ----------
+    global_offsets : torch.Tensor
+        CSC offsets, can live on the CPU
+    global_indices : torch.Tensor
+        CSC indices, can live on the CPU
+    partition_size : int
+        number of process groups across which graph is partitioned,
+        i.e. the number of graph partitions
+    partition_rank : int
+        rank within process group managing the distributed graph, i.e.
+        the rank determining which partition the corresponding local rank
+        will manage
+    device : torch.device
+        device connected to the passed partition rank, i.e. the device
+        on which the local graph and related buffers will live on
+    """
+
+    num_global_src_nodes = global_indices.max().item() + 1
+    num_global_dst_nodes = global_offsets.size(0) - 1
+    num_dst_nodes_per_partition = (
+        num_global_dst_nodes + partition_size - 1
+    ) // partition_size
+    num_src_nodes_per_partition = (
+        num_global_src_nodes + partition_size - 1
+    ) // partition_size
+
+    mapping_dst_ids_to_ranks = (
+        torch.arange(
+            num_global_dst_nodes,
+            dtype=global_offsets.dtype,
+            device=global_offsets.device,
+        )
+        // num_dst_nodes_per_partition
+    )
+    mapping_src_ids_to_ranks = (
+        torch.arange(
+            num_global_src_nodes,
+            dtype=global_offsets.dtype,
+            device=global_offsets.device,
+        )
+        // num_src_nodes_per_partition
+    )
+
+    return partition_graph_with_id_mapping(
+        global_offsets,
+        global_indices,
+        mapping_src_ids_to_ranks,
+        mapping_dst_ids_to_ranks,
+        partition_size,
+        partition_rank,
+        device,
+    )
+
+
+def partition_graph_coordinatewise(
```

Comment:
Would prefer if this was named slightly differently to make it explicit that you're specifying coordinate ranges. Maybe something like `partition_graph_by_coordinate_bounding_box`?

Reply:
good suggestion 👍🏽